### DISCUSSION TITLE
How to return Node.js `Buffer` with TypeScript 5.7?

### DISCUSSION THREAD

**Discussion by marvinruder** [honojs/hono#3700] [Q&A]
The code

```ts
import { serve } from '@hono/node-server'
import { Hono } from 'hono'
import { etag } from 'hono/etag'

const app = new Hono()
app.use(etag())
app.get('/', (c) => c.body(Buffer.from('Hello Hono!')))
serve({ fetch: app.fetch })
```

runs as expected and shows no errors in TypeScript 5.6, but the following error is shown with TypeScript 5.7:

```
No overload matches this call.
  Overload 1 of 2, '(data: Data | null, status?: StatusCode | undefined, headers?: HeaderRecord | undefined): Response', gave the following error.
    Argument of type 'Buffer<ArrayBuffer>' is not assignable to parameter of type 'Data | null'.
      Type 'Buffer<ArrayBuffer>' is missing the following properties from type 'ArrayBuffer': maxByteLength, resizable, resize, detached, and 2 more.
  Overload 2 of 2, '(data: Data | null, init?: ResponseInit | undefined): Response', gave the following error.
    Argument of type 'Buffer<ArrayBuffer>' is not assignable to parameter of type 'Data | null'.
      Type 'Buffer<ArrayBuffer>' is missing the following properties from type 'ArrayBuffer': maxByteLength, resizable, resize, detached, and 2 more.
```

This is because a `Buffer` is no longer an `ArrayBuffer` (see https://devblogs.microsoft.com/typescript/announcing-typescript-5-7/#typedarrays-are-now-generic-over-arraybufferlike, https://github.com/microsoft/TypeScript/pull/59417 for details), and the `Data` type can only be a `string`, `ArrayBuffer`, or `ReadableStream`.

Returning the `ArrayBuffer` within the `Buffer` using

```ts
app.get('/', (c) => c.body(Buffer.from('Hello Hono!').buffer))
```

shows no type error, but leads to unexpected results, as responses are 8192 bytes long and contain lots of unrelated data. The reason for this is the shared memory pool from the `Buffer` module:

> The `Buffer` module pre-allocates an internal `Buffer` instance of size [`Buffer.poolSize`](https://nodejs.org/api/buffer.html#class-property-bufferpoolsize) that is used as a pool for the fast allocation of new `Buffer` instances created using [`Buffer.allocUnsafe()`](https://nodejs.org/api/buffer.html#static-method-bufferallocunsafesize), [`Buffer.from(array)`](https://nodejs.org/api/buffer.html#static-method-bufferfromarray), [`Buffer.from(string)`](https://nodejs.org/api/buffer.html#static-method-bufferfromstring-encoding), and [`Buffer.concat()`](https://nodejs.org/api/buffer.html#static-method-bufferconcatlist-totallength) only when `size` is less than `Buffer.poolSize >>> 1` (floor of [`Buffer.poolSize`](https://nodejs.org/api/buffer.html#class-property-bufferpoolsize) divided by two). (https://nodejs.org/api/buffer.html#static-method-bufferallocunsafesize)

What is the recommended way to use a Node.js `Buffer` as the body of a response? Do Hono’s types need to be updated to conform with the changes in TypeScript 5.7?

**Comment by Sallyfafafa** [comment]
Have you solved it yet? I'm having this problem too!

**Comment by yusukebe** [comment]
We should make this an issue in `honojs/hono` repo. I'll transfer it. Thanks!
